crates/ruff/src/rules/flake8_pyi/rules/duplicate_union_member.rs

```diff
@@ -1,89 +1,79 @@
+        // If we've already seen this union member, raise a violation.
+        if !seen_nodes.insert(expr.into()) {
+            let mut diagnostic = Diagnostic::new(
+                DuplicateUnionMember {
+                    duplicate_name: checker.generator().expr(expr),
+                },
+                expr.range(),
+            );
+            if checker.patch(diagnostic.kind.rule()) {
+                // Delete the "|" character as well as the duplicate value by reconstructing the
+                // parent without the duplicate.
 
-    // If we've already seen this union member, raise a violation.
-    if !seen_nodes.insert(expr.into()) {
-        let mut diagnostic = Diagnostic::new(
-            DuplicateUnionMember {
-                duplicate_name: checker.generator().expr(expr),
-            },
-            expr.range(),
-        );
-        if checker.patch(diagnostic.kind.rule()) {
-            // Delete the "|" character as well as the duplicate value by reconstructing the
-            // parent without the duplicate.
+                // SAFETY: impossible to have a duplicate without a `parent` node.
+                let parent = parent.expect("Parent node must exist");
 
-            // SAFETY: impossible to have a duplicate without a `parent` node.
-            let parent = parent.expect("Parent node must exist");
+                // If the parent node is not a `BinOp` we will not perform a fix
+                if let Expr::BinOp(ast::ExprBinOp { left, right, .. }) = parent {
+                    // Replace the parent with its non-duplicate child.
+                    diagnostic.set_fix(Fix::automatic(Edit::range_replacement(
+                        checker
+                            .generator()
+                            .expr(if expr == left.as_ref() { right } else { left }),
+                        parent.range(),
+                    )));
+                }
+            }
+            diagnostics.push(diagnostic);
+        }
+    };
 
-            // SAFETY: Parent node must have been a `BinOp` in order for us to have traversed it.
-            let Expr::BinOp(ast::ExprBinOp { left, right, .. }) = parent else {
-                panic!("Parent node must be a BinOp");
-            };
+    // Traverse the union, collect all diagnostic members
+    traverse_union(
+        &mut check_for_duplicate_members,
+        checker.semantic(),
+        expr,
+        None,
+    );
 
-            // Replace the parent with its non-duplicate child.
-            diagnostic.set_fix(Fix::automatic(Edit::range_replacement(
-                checker
-                    .generator()
-                    .expr(if expr == left.as_ref() { right } else { left }),
-                parent.range(),
-            )));
-        }
-        checker.diagnostics.push(diagnostic);
-    }
+    // Add all diagnostics to the checker
+    checker.diagnostics.append(&mut diagnostics);
```

Comment:
(It'd be nice to append these directly as we go rather than allocate an intermediary vector, but it's not critical. I assume that lifetimes and borrowing got in the way per Discord.)

Reply:
Yeah I tried to do this a couple ways and it seems like a pain